Drewsetski component notification

hakankaradis opened the pull request:
Reopen: #9757

## Screenshots (if appropriate):
http://recordit.co/csI5NjqBLC

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

drewsetski committed:
Implemented NotificationView

drewsetski committed:
added getVendorTransition util, react-flip-move

drewsetski committed:
notification removal fix

drewsetski committed:
removed underscores before variables and methods, using "lab " alias for requiring components

drewsetski committed:
conditional statement fix

drewsetski committed:
use generateClassName for combining styles

drewsetski committed:
replace isMounted with mounted

hakankaradis committed:
fix coffeelint error

usirin commented on client/component-lab/Notification/helpers.coffee:
```diff
@@ -0,0 +1,38 @@
+
+  if not notification.type
+    throw new Error('Notification type is required.')
+  if _.keys(types).indexOf(notification.type) is -1
```
Use `Object.keys`

usirin commented on client/component-lab/Notification/helpers.coffee:
```diff
@@ -0,0 +1,38 @@
+_ = require 'lodash'
+
+Timer = (callback, delay) ->
```
Make this a proper `class`.

usirin commented on client/component-lab/Notification/helpers.coffee:
```diff
@@ -0,0 +1,38 @@
+Timer = (callback, delay) ->
+
+  timerId = undefined
+  start = undefined
```
Is there a specific reason why `null` is not being used here?

usirin commented on client/component-lab/Notification/NotificationViewContainer.coffee:
```diff
@@ -0,0 +1,79 @@
+    className = generateClassName [
+      styles.kd_notification_list
+      stateClass
+    ]
```
use `classnames` npm module here. ref: https://www.npmjs.com/package/classnames

usirin commented on client/component-lab/Notification/NotificationViewContainer.coffee:
```diff
@@ -0,0 +1,79 @@
+          opacity : 0
+
+
+  onNotificationRemove: (uid) =>
```
remove fat arrow.

usirin replied:
@hakankaradis this is still not fixed?

usirin commented on client/component-lab/Notification/NotificationView.coffee:
```diff
@@ -0,0 +1,208 @@
+      if notification.dismissible
+        className.push styles.kd_notification_dismissible
+      if notification.primaryButtonTitle or notification.secondaryButtonTitle
+        className = className.filter (word) -> word isnt styles.kd_notification_dismissible
```
use `classnames` npm module here. ref: https://www.npmjs.com/package/classnames

usirin commented on client/component-lab/Notification/NotificationView.coffee:
```diff
@@ -0,0 +1,208 @@
+        if @autoDismissible notification
+          @notificationTimer = new Helpers.Timer =>
+            @hideNotification()
+            return
```
why explicit `return`?

usirin commented on client/component-lab/Notification/NotificationView.coffee:
```diff
@@ -0,0 +1,208 @@
+    constructor: (props) ->
+
+      super props
+      @style = { zIndex: 100 + @props.index }
```
remove this from here, and define this as `defaultProps.style`

usirin commented on client/app/lib/util/getVendorTransition.coffee:
```diff
@@ -0,0 +1,14 @@
+    OTransition : 'oTransitionEnd'
+    MozTransition : 'transitionend'
+    WebkitTransition : 'webkitTransitionEnd'
+  _.keys(transitions).forEach (transitionKey) ->
```
Use `Object.keys`

usirin reviewed (changes requested):
@drewsetski  The requested changes are mostly related to the way code is written. Once they are fixed, it looks mergeable.

drewsetski committed:
changed _.keys to Object.keys

drewsetski committed:
use classnames

drewsetski committed:
use classnames

drewsetski committed:
@styles to defaultProps

drewsetski committed:
Timer is class now

hakankaradis committed:
NotificationView: remove z-index style and fix error

hakankaradis commented:
@sinan and @usirin I have updated this pr with new commits of [this pr](https://github.com/koding/koding/pull/10420)

I have removed https://github.com/koding/koding/pull/10420/files#diff-a51377e3c5c5b024cb7f53f96603be57R28 this style line, it doesn't necessary and also fixed this line of code https://github.com/koding/koding/pull/10420/files#diff-a51377e3c5c5b024cb7f53f96603be57R118

sinan commented:
@usirin can you please finalize this so that we can start using in the app.

usirin commented:
Minor fat arrow issue, but not a blocker. Thanks @hakankaradis
